algorithms/src/snark/marlin/marlin.rs

```diff
@@ -144,16 +168,19 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: MarlinMode> MarlinSNAR
 
     fn init_sponge(
         fs_parameters: &FS::Parameters,
-        batch_size: usize,
-        circuit_commitments: &[crate::polycommit::sonic_pc::Commitment<E>],
-        inputs: &[Vec<E::Fr>],
+        inputs_and_batch_sizes: &BTreeMap<CircuitId, (usize, &[Vec<E::Fr>])>,
+        circuit_commitments: &[&Vec<crate::polycommit::sonic_pc::Commitment<E>>],
```

Comment:
```suggestion
        circuit_commitments: &[&[<crate::polycommit::sonic_pc::Commitment<E>]],
```

Reply:
actually, I can see it can even be an iterator, which will allow us to get rid of some allocations

Reply:
cef8f291b